// @netlify/build-info/lib/build-systems/build-system.js
/** A list of often used dev command names to identify if the command is a dev command */
export const NPM_DEV_SCRIPTS = ['dev', 'serve', 'develop', 'start', 'run', 'web'];
/** A list of often used build command names to identify if the command is a build command */
export const NPM_BUILD_SCRIPTS = ['build'];
export class BaseBuildTool {
    project;
    id;
    name;
    version;
    configFiles = [];
    constructor(project) {
        this.project = project;
    }
    async detect() {
        const config = await this.project.fs.findUp(this.configFiles, {
            cwd: this.project.baseDirectory,
            stopAt: this.project.root,
        });
        if (config) {
            const pkgJson = await this.project.getPackageJSON(this.project.fs.dirname(config));
            this.version = pkgJson.devDependencies?.[this.id];
            return this;
        }
    }
    /** Gets a JSON from the class information */
    toJSON() {
        return {
            id: this.id,
            name: this.name,
            version: this.version,
        };
    }
}
